R/variable-transforms.R

```diff
@@ -0,0 +1,334 @@
+#'
+#' Convenience functions to test if an abstract category (AbsCat) object is a
+#' specific type. These types are defined by the properties of the abstract
+#' category object.
+#'
+#' `is.abscat.subtotal` is `x` a subtotal insertion?
+#' `is.abscat.heading` is `x` a heading insertion?
+#' `is.abscat.category` is `x` a category?
+#'
+#' @param x an AbsCat object
+#'
+#' @return logical if the AbsCat object is
+#'
+#' @name AbsCat-type-tests
+#' @keywords internal
+NULL
+
+#' @rdname AbsCat-type-tests
+#' @export
+is.abscat.subtotal <- function (x) {
+    if (is.AbsCats(x)) {
+        return(!(is.na(funcs(x))) & funcs(x) == 'subtotal')
+    }
+
+    return(!(is.na(func(x))) & func(x) == 'subtotal')
+}
+
+#' @rdname AbsCat-type-tests
+#' @export
+is.abscat.heading <- function (x) {
+    if (is.AbsCats(x)) {
+        return(is.na(funcs(x)) & !is.na(anchors(x)))
+    }
+
+    return(is.na(func(x)) & !is.na(anchor(x)))
+}
+
+#' @rdname AbsCat-type-tests
+#' @export
+is.abscat.category <- function (x) {
+    # if the class has already been specified, use that.
+    if (!is.null(x$class)) {
+        return(x$class == "Category")
+    }
+
+    # Otherwise, check if the properties of x look like a category, since
+    # value is sometimes legitamately NA, just check id, missing, and name
+    if (is.AbsCats(x)) {
+        all <- !is.na(ids(x)) & !is.na(is.na(x)) & !is.na(names(x))
+    }  else {
+        all <- !is.na(id(x)) & !is.na(is.na(x)) & !is.na(name(x))
+    }
+
+    return(all)
+}
+
+
+# make styles based on transforms and categories
+transformStyles <- function (trans, cats) {
+    # collate categories and instertions
+    all_labs <- collateCats(trans$insertions, cats)
+
+    # make a list of styles to apply
+    styles <- lapply(all_labs, function (lab) {
+        if (is.abscat.subtotal(lab)) {
+            return(subtotalStyle)
+        } else if (is.abscat.heading(lab)) {
+            return(headingStyle)
+        } else {
+            return(NULL)
+        }
+    })
+    return(styles)
+}
+
+#' @importFrom crayon make_style italic underline
+headingStyle <- c(nonas, make_style("#546499"), underline) # blue with underline
+subtotalStyle <- c(italic, make_style("#005e46"))
+
+#' Given values from an array and transforms, calculate the insertions
+#'
+#' @param vec values to transform (a single dimension of an array)
+#' @param elements AbsCats of both categories and insertions to calculate.
+#' Generally derived from `mapInsertions()`
+#' @param var_cats the `Categories` object of the transform
+#'
+#' @return the values given in `vec`, with any insertions specified in
+#' `trans` calculated
+#' @keywords internal
+calcInsertions <- function (vec, elements, var_cats) {
+    # we always calculate insertions at the lowest dimension
+    if (length(dim(vec)) > 1) {
+        halt("Calculating varaible transforms is not implemented for dimensions ",
+             "greater than 1.")
+    }
+
+    vec_out <- vapply(elements, function (element) {
```

Comment:
This looks like a function that should be pulled out for easier readability/testing.